src/ast/pass_manager.h

```diff
@@ -1,64 +1,366 @@
+      // because of the way the template is constructed.
+      outputs = { PassContext::TypeId<std::decay_t<Return>>::type_id() };
+    }
 
-public:
-  std::string name;
+    return Pass(
+        name,
+        std::move(inputs),
+        std::move(outputs),
+        [fn](PassContext &ctx) -> Result<> {
+          // Extract a tuple of all input parameters, apply the function
+          // and set the result.
+          auto args = std::make_tuple(std::ref(ctx.get<Inputs>())...);
+          auto result = std::apply(fn, args);
+          if (!result) {
+            return result.takeError();
+          }
+          if constexpr (std::is_same_v<Return, OK>) {
+            return OK();
+          } else {
+            static_assert(
+                std::is_base_of_v<PassContext::State, std::decay_t<Return>>,
+                "return value must be derived from State<...>");
+            if constexpr (std::is_reference_v<Return>) {
+              // Add the external reference.
+              ctx.put(*result);
+            } else {
+              // Just move the result.
+              ctx.put(std::move(*result));
+            }
+            return OK();
+          }
+        });
+  }
+
+  // N.B. external construction is done via `create`.
+  using Fn = std::function<Result<>(PassContext &ctx)>;
+  Pass(std::string name,
+       std::vector<int> &&inputs,
+       std::vector<int> &&outputs,
+       Fn fn)
+      : name_(std::move(name)),
+        inputs_(std::move(inputs)),
+        outputs_(std::move(outputs)),
+        fn_(std::move(fn))
+  {
+  }
+
+  std::string name_;
+  std::vector<int> inputs_;
+  std::vector<int> outputs_;
+  Fn fn_;
 };
 
+// PassManager holds passes in sequence.
+//
+// The PassManager is responsible for verifying the inputs and outputs of the
+// passes ahead of time, so that the produced graph is sound.
 class PassManager {
 public:
-  PassManager(std::ostream &out = std::cerr) : out_(out) {};
+  // registers the given pass, adding to the current list.
+  PassManager &add(Pass &&pass);
 
-  void AddPass(Pass p);
-  [[nodiscard]] int Run(PassContext &ctx);
+  // registers all the given passes.
+  PassManager &add(std::vector<Pass> &&passes);
+
+  // foreach executes the given function on every pass.
+  //
+  // The first pass to encounter errors will result in execution stopping. All
+  // warnings from all passes are collected.
+  Result<> foreach(std::function<Result<>(const Pass &)> fn);
+
+  // put adds a pass which always pushes the given static data.
+  template <typename T>
+  PassManager &put(T &t)
+  {
+    return add(Pass::create(T::type_name(), [&t]() -> T & { return t; }));
+  }
+
+  // run all registered passes.
+  //
+  // This is a canonical example of how to use `foreach` above, in many other
+  // cases a direct call will be suitable.
+  Result<PassContext> run()
+  {
+    PassContext ctx;
+    auto err = foreach([&](auto &pass) -> Result<> {
+      if (!ctx.ok())
+        return OK(); // Skip the pass.
```

Comment:
```suggestion
        return OK(); // Previous errors diagnostics, skip the pass.
```

Reply:
Fixed. I added a richer explanation to the actual method documentation, and just a note on the if here.